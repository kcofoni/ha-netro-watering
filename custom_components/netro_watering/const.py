"""Constants for the Netro Watering integration."""

DOMAIN = "netro_watering"
MANUFACTURER = "Netro"

CONF_SERIAL_NUMBER = "serial_number"
CONF_DEVICE_NAME = "device_name"
CONF_DEVICE_TYPE = "device_type"
CONF_DEVICE_HW_VERSION = "hw_version"
CONF_DEVICE_SW_VERSION = "sw_version"

CONTROLLER_DEVICE_TYPE = "controller"
SENSOR_DEVICE_TYPE = "sensor"

NETRO_SENSOR_ID = "id"
NETRO_SENSOR_CELSIUS = "celsius"
NETRO_SENSOR_FAHRENHEIT = "fahrenheit"
NETRO_SENSOR_MOISTURE = "moisture"
NETRO_SENSOR_TEMPERATURE = NETRO_SENSOR_CELSIUS
NETRO_SENSOR_LOCAL_DATE = "local_date"
NETRO_SENSOR_LOCAL_TIME = "local_time"
NETRO_SENSOR_TIME = "time"
NETRO_SENSOR_SUNLIGHT = "sunlight"
NETRO_SENSOR_BATTERY_LEVEL = "battery_level"
NETRO_DEFAULT_SENSOR_MODEL = "Whisperer"
NETRO_SPRITE_CONTROLLER_MODEL = "Sprite/Spark"
NETRO_PIXIE_CONTROLLER_MODEL = "Pixie"
NETRO_DEFAULT_ZONE_MODEL = "Virtual zone"

NETRO_CONTROLLER_ZONENUM = "zone_num"
NETRO_CONTROLLER_STATUS = "status"
NETRO_CONTROLLER_BATTERY_LEVEL = "battery_level"
NETRO_CONTROLLER_ZONES = "zones"
NETRO_CONTROLLER_WATERING = "start_watering"
NETRO_CONTROLLER_STOP_WATERING = "stop_watering"
NETRO_ZONE_ITH = "ith"
NETRO_ZONE_ENABLED = "enabled"
NETRO_ZONE_SMART = "smart"
NETRO_ZONE_NAME = "name"
NETRO_ZONE_LAST_WATERING_STATUS = "last_watering_status"
NETRO_ZONE_LAST_WATERING_START = "last_watering_start"
NETRO_ZONE_LAST_WATERING_END = "last_watering_end"
NETRO_ZONE_LAST_WATERING_SOURCE = "last_watering_source"
NETRO_ZONE_NEXT_WATERING_STATUS = "next_watering_status"
NETRO_ZONE_NEXT_WATERING_START = "next_watering_start"
NETRO_ZONE_NEXT_WATERING_END = "next_watering_end"
NETRO_ZONE_NEXT_WATERING_SOURCE = "next_watering_source"

NETRO_METADATA_LAST_ACTIVE = "last_active"
NETRO_METADATA_TIME = "time"
NETRO_METADATA_TID = "tid"
NETRO_METADATA_VERSION = "version"
NETRO_METADATA_TOKEN_LIMIT = "token_limit"
NETRO_METADATA_TOKEN_REMAINING = "token_remaining"
NETRO_METADATA_TOKEN_RESET = "token_reset"

NETRO_STATUS_ENABLE = 1
NETRO_STATUS_DISABLE = 0
NETRO_STATUS_STANDBY = "STANDBY"
NETRO_STATUS_SETUP = "SETUP"
NETRO_STATUS_ONLINE = "ONLINE"
NETRO_STATUS_WATERING = "WATERING"
NETRO_STATUS_OFFLINE = "OFFLINE"
NETRO_STATUS_SLEEPING = "SLEEPING"
NETRO_STATUS_POWEROFF = "POWEROFF"

NETRO_SCHEDULE_EXECUTED = "EXECUTED"
NETRO_SCHEDULE_EXECUTING = "EXECUTING"
NETRO_SCHEDULE_VALID = "VALID"
NETRO_SCHEDULE_STATUS = "status"
NETRO_SCHEDULE_ZONE = "zone"
NETRO_SCHEDULE_START_TIME = "start_time"
NETRO_SCHEDULE_ID = "id"
NETRO_SCHEDULE_END_TIME = "end_time"
NETRO_SCHEDULE_LOCAL_DATE = "local_date"
NETRO_SCHEDULE_LOCAL_START_TIME = "local_start_time"
NETRO_SCHEDULE_LOCAL_END_TIME = "local_end_time"
NETRO_SCHEDULE_SOURCE = "source"
NETRO_SCHEDULE_FIX = "FIX"
NETRO_SCHEDULE_SMART = "SMART"
NETRO_SCHEDULE_MANUAL = "MANUAL"

NETRO_MOISTURE_ID = "id"
NETRO_MOISTURE_ZONE = "zone"
NETRO_MOISTURE_DATE = "date"
NETRO_MOISTURE_MOISTURE = "moisture"

EXTRA_STATE_ATTRIBUTE_SEP_LEFT = "-------------------------"
EXTRA_STATE_ATTRIBUTE_SEP_RIGHT = "-------------------------"

GLOBAL_PARAMETERS = "parameters"

# the following const should be replaced by a configuration entry
SENS_REFRESH_INTERVAL_MN = 60  # minutes
CTRL_REFRESH_INTERVAL_MN = 15  # minutes
TZ_OFFSET = "+00:00"  # natively produced in UTC (do not change)
DEFAULT_WATERING_DURATION = 30  # minutes
DEFAULT_WATERING_DELAY = 0  # minutes, should be 0, if not null this parameter is a good way for testing and then be able to cancel the watering manually
MONTHS_BEFORE_SCHEDULES = 4
MONTHS_AFTER_SCHEDULES = 2
DELAY_BEFORE_REFRESH = 5  # seconds
DEFAULT_SENSOR_VALUE_DAYS_BEFORE_TODAY = 1
NETRO_API_URL = "https://api.netrohome.com/npa/v1/"

CONF_DURATION = "watering_duration"
CONF_SENS_REFRESH_INTERVAL = "sensor_refresh_interval"
CONF_SENSOR_VALUE_DAYS_BEFORE_TODAY = "sensor_value_days_before_today"
CONF_CTRL_REFRESH_INTERVAL = "controller_refresh_interval"
CONF_MONTHS_BEFORE_SCHEDULES = "months_before_schedules"
CONF_MONTHS_AFTER_SCHEDULES = "months_after_schedules"
CONF_API_URL = "netro_api_url"
CONF_DELAY_BEFORE_REFRESH = "delay_before_refresh"
CONF_DEFAULT_WATERING_DELAY = "default_watering_delay"
CONF_SLOWDOWN_FACTOR = "slowdown_factors"

ATTR_WATERING_DURATION = "duration"
ATTR_WATERING_DELAY = "delay"
ATTR_WATERING_START_TIME = "start_time"
ATTR_ZONE_ID = "zone_id"
ATTR_MOISTURE = "moisture"
ATTR_WEATHER_DATE = "date"
ATTR_WEATHER_CONDITION = "condition"
ATTR_WEATHER_RAIN = "rain"
ATTR_WEATHER_RAIN_PROB = "rain_prob"
ATTR_WEATHER_TEMP = "temp"
ATTR_WEATHER_T_MIN = "t_min"
ATTR_WEATHER_T_MAX = "t_max"
ATTR_WEATHER_T_DEW = "t_dew"
ATTR_WEATHER_WIND_SPEED = "wind_speed"
ATTR_WEATHER_HUMIDITY = "humidity"
ATTR_WEATHER_PRESSURE = "pressure"
ATTR_CONFIG_ENTRY_ID = "config_entry_id"
