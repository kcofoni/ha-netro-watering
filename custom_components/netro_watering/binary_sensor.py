"""Support for Netro watering system."""
from __future__ import annotations

from dataclasses import dataclass
import datetime
import logging
from typing import Any

from homeassistant.components.binary_sensor import (
    BinarySensorDeviceClass,
    BinarySensorEntity,
    BinarySensorEntityDescription,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity
import homeassistant.util.dt as dt_util

from .const import (
    CONF_DEVICE_TYPE,
    CONTROLLER_DEVICE_TYPE,
    DOMAIN,
    EXTRA_STATE_ATTRIBUTE_SEP_LEFT,
    EXTRA_STATE_ATTRIBUTE_SEP_RIGHT,
)
from .coordinator import NetroControllerUpdateCoordinator

_LOGGER = logging.getLogger(__name__)


@dataclass
class NetroRequiredKeysMixin:
    """Mixin for required keys."""

    netro_name: str


@dataclass
class NetroBinarySensorEntityDescription(
    BinarySensorEntityDescription, NetroRequiredKeysMixin
):
    """Defines Netro entity description."""


NETRO_ZONE_WATERING_DESCRIPTION = NetroBinarySensorEntityDescription(
    key="iswatering",
    name="Is it watering ?",
    device_class=BinarySensorDeviceClass.RUNNING,
    translation_key="iswatering",
    netro_name="watering",
    icon="mdi:watering-can-outline",
)


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up entry for a Netro sensors."""
    if config_entry.data[CONF_DEVICE_TYPE] == CONTROLLER_DEVICE_TYPE:
        controller: NetroControllerUpdateCoordinator = hass.data[DOMAIN][
            config_entry.entry_id
        ]
        # add zone binary sensors
        async_add_entities(
            [
                NetroZone(
                    controller,
                    NETRO_ZONE_WATERING_DESCRIPTION,
                    zone_key,
                )
                for zone_key in controller.active_zones
            ]
        )


class NetroZone(
    CoordinatorEntity[NetroControllerUpdateCoordinator], BinarySensorEntity
):
    """A sensor implementation for Netro Zone device."""

    entity_description: NetroBinarySensorEntityDescription
    _attr_has_entity_name = True

    def __init__(
        self,
        coordinator: NetroControllerUpdateCoordinator,
        description: NetroBinarySensorEntityDescription,
        zone_id: int,
    ) -> None:
        """Initialize the Netro sensor."""
        super().__init__(coordinator)
        self.entity_description = description
        self.zone_id = zone_id
        self._attr_unique_id = (
            f"{coordinator.active_zones[zone_id].serial_number}-{description.key}"
        )
        self._attr_device_info = coordinator.active_zones[zone_id].device_info

    @property
    def is_on(self) -> bool | None:
        """Return True if the current zone is currently watering."""
        return getattr(
            self.coordinator.active_zones[self.zone_id],
            self.entity_description.netro_name,
        )

    @property
    def extra_state_attributes(self) -> dict[str, Any]:
        """Return state attributes."""
        zone_attributes = {
            "zone id": self.zone_id,
            "update interval": f"{round(self.coordinator.update_interval.total_seconds() / 60)} mn"
            if self.coordinator.update_interval is not None
            else None,
        }
        if self.coordinator.current_slowdown_factor > 1:
            zone_attributes[
                "slowdown factor"
            ] = self.coordinator.current_slowdown_factor
        meta_attributes = {
            EXTRA_STATE_ATTRIBUTE_SEP_LEFT: EXTRA_STATE_ATTRIBUTE_SEP_RIGHT,
            "request time (UTC)": self.coordinator.metadata.time
            if self.coordinator.metadata is not None
            else None,
            "last active date": dt_util.as_local(
                self.coordinator.metadata.last_active_date.replace(
                    tzinfo=datetime.UTC
                )
            )
            if self.coordinator.metadata is not None
            else None,
            "transaction id": self.coordinator.metadata.tid
            if self.coordinator.metadata is not None
            else None,
            "token limit": self.coordinator.metadata.token_limit
            if self.coordinator.metadata is not None
            else None,
            "token remaining": self.coordinator.metadata.token_remaining
            if self.coordinator.metadata is not None
            else None,
            "token reset": dt_util.as_local(
                self.coordinator.metadata.token_reset_date.replace(
                    tzinfo=datetime.UTC
                )
            )
            if self.coordinator.metadata is not None
            else None,
        }
        return zone_attributes | meta_attributes
