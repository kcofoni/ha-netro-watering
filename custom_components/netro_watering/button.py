"""Support du bouton Refresh pour Netro Watering."""

from __future__ import annotations

import logging

from homeassistant.components.button import ButtonEntity, ButtonEntityDescription
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import DOMAIN
from .coordinator import NetroControllerUpdateCoordinator

_LOGGER = logging.getLogger(__name__)

NETRO_CONTROLLER_BUTTON_DESCRIPTION = ButtonEntityDescription(
    key="refresh",
    name="Refresh",
    entity_registry_enabled_default=True,
    translation_key="refresh",
    icon="mdi:refresh",
)


async def async_setup_entry(
    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
) -> None:
    """Set up the button platform."""
    controller: NetroControllerUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
    _LOGGER.info("Adding button entity for controller = %s", controller.device_name)

    async_add_entities(
        [NetroRefreshButton(controller, NETRO_CONTROLLER_BUTTON_DESCRIPTION)]
    )


class NetroRefreshButton(
    CoordinatorEntity[NetroControllerUpdateCoordinator], ButtonEntity
):
    """Bouton qui force un refresh du contrôleur Netro."""

    entity_description: ButtonEntityDescription
    _attr_has_entity_name = True

    def __init__(
        self,
        coordinator: NetroControllerUpdateCoordinator,
        description: ButtonEntityDescription,
    ) -> None:
        """Initialize the Netro refresh button entity.

        Args:
            coordinator: The data update coordinator for Netro.
            description: The entity description for the button.
        """
        super().__init__(coordinator)
        self.entity_description = description
        self._attr_unique_id = f"{coordinator.serial_number}-{description.key}"
        self._attr_device_info = coordinator.device_info

    async def async_press(self) -> None:
        """Pressing the button requests an immediate update."""
        _LOGGER.debug(
            "Netro: refresh button pressed for %s", self.coordinator.serial_number
        )

        # Si ton coordinator expose une méthode spécifique (ex: refresh_controller()),
        # tu peux l'appeler ici avant/à la place du request_refresh.
        # await self.coordinator.refresh_controller()

        # Dans tous les cas, on (re)planifie une mise à jour via le DataUpdateCoordinator
        await self.coordinator.async_request_refresh()
