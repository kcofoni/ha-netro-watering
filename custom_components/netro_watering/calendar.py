"""Support for Netro Watering system."""
from __future__ import annotations

import datetime
import logging

from homeassistant.components.calendar import CalendarEntity, CalendarEvent
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity import EntityDescription
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import CONF_DEVICE_TYPE, CONTROLLER_DEVICE_TYPE, DOMAIN
from .coordinator import NetroControllerUpdateCoordinator

_LOGGER = logging.getLogger(__name__)

NETRO_CALENDAR_DESCRIPTION = EntityDescription(
    key="schedules",
    name="Schedules",
    entity_registry_enabled_default=True,
    translation_key="schedules",
    icon="mdi:calendar-clock",
)


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up entry for the controller calendar."""
    if config_entry.data[CONF_DEVICE_TYPE] == CONTROLLER_DEVICE_TYPE:
        _LOGGER.info("Adding calendar entity")
        # add controller calendar
        async_add_entities(
            [
                NetroCalendar(
                    hass.data[DOMAIN][config_entry.entry_id],
                    NETRO_CALENDAR_DESCRIPTION,
                )
            ]
        )


class NetroCalendar(
    CoordinatorEntity[NetroControllerUpdateCoordinator], CalendarEntity
):
    """A calendar implementation for Netro Controller device."""

    _attr_has_entity_name = True

    def __init__(
        self,
        coordinator: NetroControllerUpdateCoordinator,
        description: EntityDescription,
    ) -> None:
        """Initialize the Netro calendar."""
        super().__init__(coordinator)
        self.entity_description = description
        self._attr_unique_id = f"{coordinator.serial_number}-{description.key}"
        self._attr_device_info = coordinator.device_info

    @property
    def event(self) -> CalendarEvent | None:
        """Return current or next upcoming schedule if any."""
        schedule = self.coordinator.current_calendar_schedule
        return (
            CalendarEvent(
                schedule["start"],
                schedule["end"],
                schedule["summary"],
                schedule["description"],
            )
            if schedule is not None
            else None
        )

    async def async_get_events(
        self,
        hass: HomeAssistant,
        start_date: datetime.datetime,
        end_date: datetime.datetime,
    ) -> list[CalendarEvent]:
        """Get all schedules in a specific time frame."""
        return [
            CalendarEvent(
                schedule["start"],
                schedule["end"],
                schedule["summary"],
                schedule["description"],
            )
            for schedule in self.coordinator.calendar_schedules(start_date, end_date)
        ]
