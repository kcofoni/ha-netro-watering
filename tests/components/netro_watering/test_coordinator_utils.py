"""Tests for coordinator utility functions."""

import datetime

import pytest

from custom_components.netro_watering.coordinator import (
    Meta,
    get_slowdown_factor,
    prepare_slowdown_factors,
)


class TestPrepareSlowdownFactors:
    """Test suite for prepare_slowdown_factors function."""

    def test_prepare_slowdown_factors_basic_case(self, snapshot):
        """Test basic case with simple time windows."""
        input_data = [
            {"from": "08:30", "to": "12:00", "sdf": 2},
            {"from": "14:15:30", "to": "18:45", "sdf": 3},
            {"from": "22:00", "to": "06:30", "sdf": 4},  # Crosses midnight
        ]

        result = prepare_slowdown_factors(input_data)

        # Snapshot test - will create reference on first run
        assert result == snapshot

    def test_prepare_slowdown_factors_empty_list(self, snapshot):
        """Test with empty list."""
        input_data = []

        result = prepare_slowdown_factors(input_data)

        assert result == snapshot

    def test_prepare_slowdown_factors_none_input(self, snapshot):
        """Test with None input."""
        result = prepare_slowdown_factors(None)

        assert result == snapshot

    def test_prepare_slowdown_factors_complex_case(self, snapshot):
        """Test with complex time windows including various formats."""
        input_data = [
            {"from": "00:00", "to": "06:00", "sdf": 1},
            {"from": "06:30:15", "to": "09:45:30", "sdf": 2},
            {"from": "12:00:00", "to": "14:30:45", "sdf": 3},
            {"from": "18:15", "to": "23:59:59", "sdf": 4},
            {"from": "23:30", "to": "02:15", "sdf": 5},  # Crosses midnight
        ]

        result = prepare_slowdown_factors(input_data)

        assert result == snapshot

    def test_prepare_slowdown_factors_midnight_crossing(self, snapshot):
        """Test specifically for midnight crossing scenarios."""
        input_data = [
            {"from": "22:30", "to": "06:30", "sdf": 2},
            {"from": "23:45:30", "to": "01:15:45", "sdf": 3},
        ]

        result = prepare_slowdown_factors(input_data)

        assert result == snapshot

    def test_prepare_slowdown_factors_with_seconds(self, snapshot):
        """Test with time formats including seconds."""
        input_data = [
            {"from": "08:30:15", "to": "12:45:30", "sdf": 2},
            {"from": "14:00:00", "to": "18:30:59", "sdf": 3},
        ]

        result = prepare_slowdown_factors(input_data)

        assert result == snapshot


class TestGetSlowdownFactor:
    """Test suite for get_slowdown_factor function."""

    def test_get_slowdown_factor_default_no_factors(self, snapshot):
        """Test that default factor 1 is returned when no slowdown factors are provided."""
        test_time = datetime.time(10, 30, 0)

        result = get_slowdown_factor(None, test_time)

        assert result == snapshot

    def test_get_slowdown_factor_empty_list(self, snapshot):
        """Test that default factor 1 is returned when empty list is provided."""
        test_time = datetime.time(10, 30, 0)

        result = get_slowdown_factor([], test_time)

        assert result == snapshot

    def test_get_slowdown_factor_time_in_window(self, snapshot):
        """Test factor is returned when time falls within a defined window."""
        slowdown_factors = [
            {"from": 8.5, "to": 12.0, "sdf": 2},  # 08:30 - 12:00
            {"from": 14.0, "to": 18.0, "sdf": 3},  # 14:00 - 18:00
        ]

        # Time within first window (10:15 = 10.25)
        test_time = datetime.time(10, 15, 0)
        result = get_slowdown_factor(slowdown_factors, test_time)

        assert result == snapshot

    def test_get_slowdown_factor_time_outside_windows(self, snapshot):
        """Test default factor 1 is returned when time is outside all windows."""
        slowdown_factors = [
            {"from": 8.5, "to": 12.0, "sdf": 2},  # 08:30 - 12:00
            {"from": 14.0, "to": 18.0, "sdf": 3},  # 14:00 - 18:00
        ]

        # Time outside windows (13:00 = 13.0)
        test_time = datetime.time(13, 0, 0)
        result = get_slowdown_factor(slowdown_factors, test_time)

        assert result == snapshot

    def test_get_slowdown_factor_exact_boundary_start(self, snapshot):
        """Test factor is returned when time is exactly at window start."""
        slowdown_factors = [
            {"from": 8.5, "to": 12.0, "sdf": 2},  # 08:30 - 12:00
        ]

        # Time exactly at start (08:30 = 8.5)
        test_time = datetime.time(8, 30, 0)
        result = get_slowdown_factor(slowdown_factors, test_time)

        assert result == snapshot

    def test_get_slowdown_factor_exact_boundary_end(self, snapshot):
        """Test factor is returned when time is exactly at window end."""
        slowdown_factors = [
            {"from": 8.5, "to": 12.0, "sdf": 2},  # 08:30 - 12:00
        ]

        # Time exactly at end (12:00 = 12.0)
        test_time = datetime.time(12, 0, 0)
        result = get_slowdown_factor(slowdown_factors, test_time)

        assert result == snapshot

    def test_get_slowdown_factor_midnight_crossing_positive(self, snapshot):
        """Test factor is returned for midnight crossing window (positive time)."""
        slowdown_factors = [
            {"from": -2.0, "to": 6.5, "sdf": 4},  # 22:00 - 06:30 (crosses midnight)
        ]

        # Time in positive part (01:00 = 1.0)
        test_time = datetime.time(1, 0, 0)
        result = get_slowdown_factor(slowdown_factors, test_time)

        assert result == snapshot

    def test_get_slowdown_factor_midnight_crossing_negative(self, snapshot):
        """Test factor is returned for midnight crossing window (negative time)."""
        slowdown_factors = [
            {"from": -2.0, "to": 6.5, "sdf": 4},  # 22:00 - 06:30 (crosses midnight)
        ]

        # Time in negative part (23:00 = 23.0, converts to -1.0)
        test_time = datetime.time(23, 0, 0)
        result = get_slowdown_factor(slowdown_factors, test_time)

        assert result == snapshot

    def test_get_slowdown_factor_with_seconds(self, snapshot):
        """Test factor calculation with seconds precision."""
        slowdown_factors = [
            {
                "from": 8.504166666666666,
                "to": 12.758333333333333,
                "sdf": 2,
            },  # 08:30:15 - 12:45:30
        ]

        # Time with seconds (08:45:30 = 8.758333...)
        test_time = datetime.time(8, 45, 30)
        result = get_slowdown_factor(slowdown_factors, test_time)

        assert result == snapshot

    def test_get_slowdown_factor_first_match_wins(self, snapshot):
        """Test that first matching window is used when multiple windows overlap."""
        slowdown_factors = [
            {"from": 8.0, "to": 12.0, "sdf": 2},  # 08:00 - 12:00
            {"from": 10.0, "to": 14.0, "sdf": 3},  # 10:00 - 14:00 (overlaps)
        ]

        # Time in both windows (11:00 = 11.0) - should return first match (sdf=2)
        test_time = datetime.time(11, 0, 0)
        result = get_slowdown_factor(slowdown_factors, test_time)

        assert result == snapshot


class TestMeta:
    """Test suite for Meta class."""

    def test_meta_initialization_valid_data(self, snapshot):
        """Test Meta initialization with valid ISO datetime strings."""
        meta = Meta(
            last_active="2024-10-10T08:30:15",
            time="2024-10-10T10:45:30",
            tid="TID123456",
            version="1.2.3",
            token_limit=1000,
            token_remaining=750,
            token_reset="2024-10-11T00:00:00",
        )

        result = {
            "version": meta.version,
            "token_limit": meta.token_limit,
            "token_remaining": meta.token_remaining,
            "tid": meta.tid,
            "last_active_date": meta.last_active_date.isoformat(),
            "time": meta.time.isoformat(),
            "token_reset_date": meta.token_reset_date.isoformat(),
        }

        assert result == snapshot

    def test_meta_datetime_conversion_with_microseconds(self, snapshot):
        """Test Meta handles ISO datetime strings with microseconds."""
        meta = Meta(
            last_active="2024-10-10T08:30:15.123456",
            time="2024-10-10T10:45:30.987654",
            tid="TID789",
            version="2.0.0",
            token_limit=500,
            token_remaining=250,
            token_reset="2024-10-11T00:00:00.000000",
        )

        result = {
            "last_active_microseconds": meta.last_active_date.microsecond,
            "time_microseconds": meta.time.microsecond,
            "token_reset_microseconds": meta.token_reset_date.microsecond,
        }

        assert result == snapshot

    def test_meta_datetime_conversion_with_timezone(self, snapshot):
        """Test Meta handles ISO datetime strings with timezone info."""
        meta = Meta(
            last_active="2024-10-10T08:30:15+02:00",
            time="2024-10-10T10:45:30Z",
            tid="TID_TZ",
            version="1.0.0",
            token_limit=2000,
            token_remaining=1500,
            token_reset="2024-10-11T00:00:00-05:00",
        )

        result = {
            "last_active_tz": str(meta.last_active_date.tzinfo),
            "time_tz": str(meta.time.tzinfo),
            "token_reset_tz": str(meta.token_reset_date.tzinfo),
            "last_active_utc_offset": meta.last_active_date.utcoffset(),
            "time_utc_offset": meta.time.utcoffset(),
        }

        assert result == snapshot

    def test_meta_edge_case_midnight(self, snapshot):
        """Test Meta with midnight times."""
        meta = Meta(
            last_active="2024-10-10T00:00:00",
            time="2024-10-10T00:00:00",
            tid="MIDNIGHT",
            version="0.0.1",
            token_limit=100,
            token_remaining=0,
            token_reset="2024-10-11T00:00:00",
        )

        result = {
            "hour": meta.last_active_date.hour,
            "minute": meta.time.minute,
            "second": meta.token_reset_date.second,
            "token_remaining_zero": meta.token_remaining,
        }

        assert result == snapshot

    def test_meta_invalid_datetime_format(self, snapshot):
        """Test Meta raises ValueError for invalid datetime format."""
        with pytest.raises(ValueError) as exc_info:
            Meta(
                last_active="invalid-date-format",
                time="2024-10-10T10:45:30",
                tid="ERROR_TEST",
                version="1.0.0",
                token_limit=1000,
                token_remaining=500,
                token_reset="2024-10-11T00:00:00",
            )

        result = {
            "exception_type": type(exc_info.value).__name__,
            "exception_message_contains_invalid": "Invalid isoformat string"
            in str(exc_info.value),
        }

        assert result == snapshot

    def test_meta_invalid_time_format(self, snapshot):
        """Test Meta raises ValueError for invalid time format."""
        with pytest.raises(ValueError) as exc_info:
            Meta(
                last_active="2024-10-10T08:30:15",
                time="not-a-time",
                tid="TIME_ERROR",
                version="1.0.0",
                token_limit=1000,
                token_remaining=500,
                token_reset="2024-10-11T00:00:00",
            )

        result = {
            "exception_type": type(exc_info.value).__name__,
            "has_error_message": len(str(exc_info.value)) > 0,
        }

        assert result == snapshot

    def test_meta_invalid_token_reset_format(self, snapshot):
        """Test Meta raises ValueError for invalid token_reset format."""
        with pytest.raises(ValueError) as exc_info:
            Meta(
                last_active="2024-10-10T08:30:15",
                time="2024-10-10T10:45:30",
                tid="RESET_ERROR",
                version="1.0.0",
                token_limit=1000,
                token_remaining=500,
                token_reset="bad-reset-time",
            )

        result = {
            "exception_type": type(exc_info.value).__name__,
            "exception_occurred": True,
        }

        assert result == snapshot

    def test_meta_extreme_values(self, snapshot):
        """Test Meta with extreme but valid values."""
        meta = Meta(
            last_active="1970-01-01T00:00:00",  # Unix epoch
            time="2099-12-31T23:59:59",  # Far future
            tid="EXTREME_TEST_" + "X" * 100,  # Long TID
            version="999.999.999",
            token_limit=999999,
            token_remaining=0,
            token_reset="2024-12-31T23:59:59.999999",
        )

        result = {
            "last_active_year": meta.last_active_date.year,
            "time_year": meta.time.year,
            "tid_length": len(meta.tid),
            "version": meta.version,
            "token_limit": meta.token_limit,
            "token_remaining": meta.token_remaining,
            "token_reset_microsecond": meta.token_reset_date.microsecond,
        }

        assert result == snapshot
