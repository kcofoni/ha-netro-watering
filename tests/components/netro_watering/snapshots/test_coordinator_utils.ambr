# serializer version: 1
# name: TestGetSlowdownFactor.test_get_slowdown_factor_default_no_factors
  1
# ---
# name: TestGetSlowdownFactor.test_get_slowdown_factor_empty_list
  1
# ---
# name: TestGetSlowdownFactor.test_get_slowdown_factor_exact_boundary_end
  2
# ---
# name: TestGetSlowdownFactor.test_get_slowdown_factor_exact_boundary_start
  2
# ---
# name: TestGetSlowdownFactor.test_get_slowdown_factor_first_match_wins
  2
# ---
# name: TestGetSlowdownFactor.test_get_slowdown_factor_midnight_crossing_negative
  4
# ---
# name: TestGetSlowdownFactor.test_get_slowdown_factor_midnight_crossing_positive
  4
# ---
# name: TestGetSlowdownFactor.test_get_slowdown_factor_time_in_window
  2
# ---
# name: TestGetSlowdownFactor.test_get_slowdown_factor_time_outside_windows
  1
# ---
# name: TestGetSlowdownFactor.test_get_slowdown_factor_with_seconds
  2
# ---
# name: TestMeta.test_meta_datetime_conversion_with_microseconds
  dict({
    'last_active_microseconds': 123456,
    'time_microseconds': 987654,
    'token_reset_microseconds': 0,
  })
# ---
# name: TestMeta.test_meta_datetime_conversion_with_timezone
  dict({
    'last_active_tz': 'UTC+02:00',
    'last_active_utc_offset': datetime.timedelta(seconds=7200),
    'time_tz': 'UTC',
    'time_utc_offset': datetime.timedelta(0),
    'token_reset_tz': 'UTC-05:00',
  })
# ---
# name: TestMeta.test_meta_edge_case_midnight
  dict({
    'hour': 0,
    'minute': 0,
    'second': 0,
    'token_remaining_zero': 0,
  })
# ---
# name: TestMeta.test_meta_extreme_values
  dict({
    'last_active_year': 1970,
    'tid_length': 113,
    'time_year': 2099,
    'token_limit': 999999,
    'token_remaining': 0,
    'token_reset_microsecond': 999999,
    'version': '999.999.999',
  })
# ---
# name: TestMeta.test_meta_initialization_valid_data
  dict({
    'last_active_date': '2024-10-10T08:30:15',
    'tid': 'TID123456',
    'time': '2024-10-10T10:45:30',
    'token_limit': 1000,
    'token_remaining': 750,
    'token_reset_date': '2024-10-11T00:00:00',
    'version': '1.2.3',
  })
# ---
# name: TestMeta.test_meta_invalid_datetime_format
  dict({
    'exception_message_contains_invalid': True,
    'exception_type': 'ValueError',
  })
# ---
# name: TestMeta.test_meta_invalid_time_format
  dict({
    'exception_type': 'ValueError',
    'has_error_message': True,
  })
# ---
# name: TestMeta.test_meta_invalid_token_reset_format
  dict({
    'exception_occurred': True,
    'exception_type': 'ValueError',
  })
# ---
# name: TestPrepareSlowdownFactors.test_prepare_slowdown_factors_basic_case
  list([
    dict({
      'from': 8.5,
      'sdf': 2,
      'to': 12.0,
    }),
    dict({
      'from': 14.258333333333333,
      'sdf': 3,
      'to': 18.75,
    }),
    dict({
      'from': -2.0,
      'sdf': 4,
      'to': 6.5,
    }),
  ])
# ---
# name: TestPrepareSlowdownFactors.test_prepare_slowdown_factors_complex_case
  list([
    dict({
      'from': 0.0,
      'sdf': 1,
      'to': 6.0,
    }),
    dict({
      'from': 6.504166666666666,
      'sdf': 2,
      'to': 9.758333333333333,
    }),
    dict({
      'from': 12.0,
      'sdf': 3,
      'to': 14.5125,
    }),
    dict({
      'from': 18.25,
      'sdf': 4,
      'to': 23.999722222222225,
    }),
    dict({
      'from': -0.5,
      'sdf': 5,
      'to': 2.25,
    }),
  ])
# ---
# name: TestPrepareSlowdownFactors.test_prepare_slowdown_factors_empty_list
  list([
  ])
# ---
# name: TestPrepareSlowdownFactors.test_prepare_slowdown_factors_midnight_crossing
  list([
    dict({
      'from': -1.5,
      'sdf': 2,
      'to': 6.5,
    }),
    dict({
      'from': -0.24166666666666714,
      'sdf': 3,
      'to': 1.2625,
    }),
  ])
# ---
# name: TestPrepareSlowdownFactors.test_prepare_slowdown_factors_none_input
  None
# ---
# name: TestPrepareSlowdownFactors.test_prepare_slowdown_factors_with_seconds
  list([
    dict({
      'from': 8.504166666666666,
      'sdf': 2,
      'to': 12.758333333333333,
    }),
    dict({
      'from': 14.0,
      'sdf': 3,
      'to': 18.51638888888889,
    }),
  ])
# ---
